source/extensions/filters/http/response_map/response_map_filter.cc

```diff
@@ -0,0 +1,254 @@
+#include "extensions/filters/http/response_map/response_map_filter.h"
+
+#include "envoy/http/codes.h"
+#include "envoy/http/header_map.h"
+
+#include "common/common/empty_string.h"
+#include "common/common/enum_to_int.h"
+#include "common/common/logger.h"
+#include "common/http/header_map_impl.h"
+#include "common/http/headers.h"
+#include "common/http/utility.h"
+
+#include "extensions/filters/http/well_known_names.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace ResponseMapFilter {
+
+ResponseMapFilterConfig::ResponseMapFilterConfig(
+    const envoy::extensions::filters::http::response_map::v3::ResponseMap& proto_config,
+    const std::string&, Server::Configuration::FactoryContext& context)
+    : response_map_(ResponseMap::Factory::create(proto_config.response_map(), context,
+                                                 context.messageValidationVisitor())) {}
+
+FilterConfigPerRoute::FilterConfigPerRoute(
+    const envoy::extensions::filters::http::response_map::v3::ResponseMapPerRoute& proto_config,
+    Server::Configuration::ServerFactoryContext& context,
+    ProtobufMessage::ValidationVisitor& validationVisitor)
+    : disabled_(proto_config.disabled()),
+      response_map_(proto_config.has_response_map()
+                        ? ResponseMap::Factory::create(proto_config.response_map(), context,
+                                                       validationVisitor)
+                        : nullptr) {}
+
+ResponseMapFilter::ResponseMapFilter(ResponseMapFilterConfigSharedPtr config) : config_(config) {}
+
+/*
+ * Get FilterConfigPerRoute if one exists for the current route.
+ */
+const FilterConfigPerRoute* ResponseMapFilter::getRouteSpecificConfig(void) {
+  const auto& route = encoder_callbacks_->route();
+  if (route == nullptr) {
+    return nullptr;
+  }
+
+  const auto* per_route_config =
+      Http::Utility::resolveMostSpecificPerFilterConfig<FilterConfigPerRoute>(
+          Extensions::HttpFilters::HttpFilterNames::get().ResponseMap, route);
+  ENVOY_LOG(trace, "response map filter: found route. has per_route_config? {}",
+            per_route_config != nullptr);
+
+  return per_route_config;
+}
+
+/*
+ * Called if there are any downstream headers to decode from the client or a previous filter.
+ *
+ * Not guaranteed to be called. In particular, if Envoy sends a local reply with HTTP 426
+ * as an upgrade response for HTTP/1.0 requests, we may never decode any of the client's
+ * original headers, because the request was rejected at the initial HTTP/1.0 line.
```

Comment:
This comment might be better off above the original `decodeHeaders` function declaration.